### DISCUSSION TITLE
Planewave in 3D model

### DISCUSSION THREAD

**Discussion by Din-Y** [NanoComp/meep#2589] [General]
Hello everyone, 
I am still trying to realize a 3D model with planewave at oblique incidence. For oblique incidence I use theta and phi (azimuth) angles and two dft monitors to get Ey and Ez fields. At phi = 45 degree I should get Ey and Ez field with the same amplitudes and the assumption is confirmed by lumerical results. But MEEP performed one component field (which is specified in source) with high amplitude and the other component field with low amplitude. Could you tell me please, how I can solve the problem? Thanks!
![m](https://github.com/NanoComp/meep/assets/132554077/f048a6b4-2b30-4ed3-a814-355bd42e7f70)

**Comment by stevengj** [comment]
What source current are you using?

**Comment by Din-Y** [comment]
> What source current are you using?

The images obtained with GaussianSource, additionally I tried ContinuousSource and the results were practically the same.

**Comment by stevengj** [comment]
The time dependence doesn't matter here since you are using a DFT monitor (though I would recommend a GaussianSource so that you only need a short simulation).  The key question is what is the spatial profile and orientation of the current source?

Can you just post the code for your current source?

(I'm transferring this to a discussion since it seems unlikely to be a bug in Meep, rather than just a misunderstanding of how sources work.)

**Reply by Din-Y** [reply]
```
import cmath
import math

import numpy as np

import meep as mp
import matplotlib.pyplot as plt
resolution = 12  # pixels/μm

dpml = 1.0  # PML thickness
dsub = 3.0  # substrate thickness
dpad = 3.0  # length of padding between grating and PML
gp = 10.0  # grating period
gh = 0.5  # grating height
gdc = 0.5  # grating duty cycle

sx = dpml + dsub + gh + dpad + dpml
sy = gp
sz = gp

cell_size = mp.Vector3(sx, sy, sz)
pml_layers = [mp.PML(thickness=dpml, direction=mp.X)]

wvl = 0.5  # center wavelength
fcen = 1 / wvl  # center frequency
df = 0.05 * fcen  # frequency width

ng = 1.5
n = 1      #air
glass = mp.Medium(index=ng)
air = mp.Medium(index=n)

theta_in = math.radians(10.7)
phi = math.radians(45)

k = mp.Vector3(fcen * ng).rotate(mp.Vector3(z=1), theta_in).rotate(mp.Vector3(x=1), phi)

if theta_in == 0:
    k = mp.Vector3(0, 0, 0)

def pw_amp(k, x0):
    def _pw_amp(x):
        return cmath.exp(1j * 2 * math.pi * k.dot(x + x0))
    return _pw_amp

src_pt = mp.Vector3(-0.5 * sx + dpml + 0.3 * dsub, 0, 0)
sources = [
    mp.Source(
        mp.GaussianSource(fcen, fwidth=df),
        component=mp.Ez,
        center=src_pt,
        size=mp.Vector3(0, sy, sz),
        amp_func=pw_amp(k, src_pt),
    )
]

sim = mp.Simulation(
    resolution=resolution,
    cell_size=cell_size,
    boundary_layers=pml_layers,
    k_point=k,
    default_material=glass,
    sources=sources,
)

tran_pt = mp.Vector3(0.5 * sx - dpml - 0.5 * dpad, 0, 0)

dft_input = sim.add_dft_fields([mp.Ez], fcen, 0, 1, center=mp.Vector3(-0.5 * sx + dpml + 0.3 * dsub, 0, 0), size=mp.Vector3(0,sy,sz))
dft_input2 = sim.add_dft_fields([mp.Ey], fcen, 0, 1, center=mp.Vector3(-0.5 * sx + dpml + 0.3 * dsub, 0, 0), size=mp.Vector3(0,sy,sz))

sim.run(until=250)
    
field_input1 = sim.get_dft_array(dft_input, mp.Ez, 0)
field_input2 = sim.get_dft_array(dft_input2, mp.Ey, 0)

plt.subplot(2,2,1)
plt.imshow(np.real(field_input1))
plt.title("Ez field", fontsize = 9)
plt.colorbar()
plt.xticks(fontsize = 0)
plt.yticks(fontsize = 0)

plt.subplot(2,2,2)
plt.imshow(np.real(field_input2))
plt.title("Ey field", fontsize = 9)
plt.colorbar()
plt.xticks(fontsize = 0)
plt.yticks(fontsize = 0)
```

**Comment by stevengj** [comment]
So you have an electric current (J) sheet in a yz plane, with the current oriented in the z direction (Ez) and an amplitude ~exp(ikx) determined by your desired planewave angle.  Unsurprisingly, this produces a planewave that is polarized more in the z direction than in the y direction; it's certainly not purely s or p polarized.

You should read the review article https://arxiv.org/abs/1301.5366 to learn about the Equivalence Principle and how current sources relate to waves.

If you want to excite an s- or p-polarized planewave with a given angle and amplitude, it is probably easier to use a [`DiffractedPlanewave`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#diffractedplanewave) object as the `eig_band` parameter of an [`EigenmodeSource`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#eigenmodesource).

**Reply by Din-Y** [reply]
Thanks!
